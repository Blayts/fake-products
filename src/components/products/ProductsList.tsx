import { useCallback, useEffect, useRef, useState } from 'react';
import type { UIEvent } from 'react';
import { useNavigate, useLocation, useParams } from 'react-router';
import { Alert, List } from 'antd';
import { ProductItem } from './ProductItem';
import { LIMIT_PAGES } from '../../constants/base';
import { useProducts } from '../../hooks/useProducts';
import { debounce } from '../../utils';
import './ProductsList.css';

export function ProductsList() {
    const { category } = useParams();
    const navigate = useNavigate();
    const { pathname } = useLocation();
    
    const [scrollValue, setScrollValue] = useState(1);
    const [page, setPage] = useState(1);
    const { products, loading } = useProducts(category, page);
    const wrapperRef = useRef<HTMLDivElement>(null);

    const changePage = useCallback(
        debounce(() => setPage((page) => page + 1), 500),
        []
    );

    function handleClick(id: number) {
        navigate(pathname + '/product/' + id);
    }

    function handleScroll(e: UIEvent<HTMLDivElement>) {
        if (loading || page >= LIMIT_PAGES) {
            return;
        }

        const div = e.target as HTMLDivElement;
        const toBottom = div.scrollHeight - div.scrollTop - div.scrollWidth;

        if (scrollValue !== toBottom && toBottom < 300) {
            setScrollValue(toBottom);
            changePage();
        }
    }

    useEffect(() => {
        setPage(1);
        wrapperRef.current?.scrollTo?.({ top: 0 });
    }, [pathname]);

    return (
        <div className="products-wrapper" onScroll={handleScroll} ref={ wrapperRef }>
            <List loading={loading}>
                {products.map((product) => (
                    <List.Item
                        key={product.id}
                    >
                        <ProductItem onClick={ () => handleClick(product.id) } product={product}></ProductItem>
                    </List.Item>
                ))}
            </List>
            {page === LIMIT_PAGES && (
                <Alert
                    message="The demo version limit has been reached"
                    showIcon
                    type="warning"
                ></Alert>
            )}
        </div>
    );
}
