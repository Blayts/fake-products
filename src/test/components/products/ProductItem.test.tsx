import { describe, expect, test } from 'vitest';
import { render, screen } from '@testing-library/react';
import { products } from '../../mock-data/products';
import { ProductItem } from '../../../components/products/ProductItem';

describe('Render product item', () => {    
    test('Snapshot test', async () => {        
        const component = render(
            <ProductItem onClick={() => {}} product={products[0]}></ProductItem>
        );

        await screen.findByText(products[0].title);

        expect(component.container).toMatchSnapshot();
    })
})