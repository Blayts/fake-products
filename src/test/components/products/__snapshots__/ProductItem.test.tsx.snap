// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Render product item > Snapshot test 1`] = `
<div>
  <div
    class="ant-card ant-card-bordered ant-card-hoverable css-dev-only-do-not-override-7t2xvq"
    style="width: 50%;"
  >
    <div
      class="ant-card-head"
    >
      <div
        class="ant-card-head-wrapper"
      >
        <div
          class="ant-card-head-title"
        >
          Title #1
        </div>
      </div>
    </div>
    <div
      class="ant-card-body"
    >
      <div
        class="ant-row css-dev-only-do-not-override-7t2xvq"
      >
        <div
          class="ant-col ant-col-8 css-dev-only-do-not-override-7t2xvq"
        >
          <div
            class="ant-row css-dev-only-do-not-override-7t2xvq"
          >
            <img
              src="Image #1"
              style="width: 100%;"
            />
          </div>
        </div>
        <div
          class="ant-col ant-col-15 ant-col-offset-1 css-dev-only-do-not-override-7t2xvq"
          style="padding: 20px;"
        >
          <div
            class="ant-row css-dev-only-do-not-override-7t2xvq"
          >
            <b>
              Model: 
              Model #1
            </b>
          </div>
          <div
            class="ant-row css-dev-only-do-not-override-7t2xvq"
          >
            <b>
              Color: 
              Color #1
            </b>
          </div>
          <div
            class="ant-row css-dev-only-do-not-override-7t2xvq"
          >
            <b>
              Price: 
              100
              cu
            </b>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
